text/0000-dynamic-attributes.md

```diff
@@ -0,0 +1,150 @@
+
+One such usecase is `lwc:dynamic` which allows component authors to lazily instantiate components. However,
+the props that we pass to the component can't be dynamic: we need to pass them in the template HTML itself.
+
+This is a problem when the component constructor is changed at runtime: the new constructor
+might need a different set of properties than the earlier one.
+
+We can, of course, not pass any props and once the element renders, query the
+element and set the props programmatically. We lose reactivity this way and also force
+the component to handle `null` props which might not be desirable.
+
+This proposal allows not just `lwc:dynamic` but all other elements to accept an object which is bound as props
+at runtime.
+
+### Prior Art
+- React ([JSX Spread Operator](https://reactjs.org/docs/jsx-in-depth.html#spread-attributes))
+- Vue ([`v-bind`](https://v3.vuejs.org/api/directives.html#v-bind))
+- Svelte ([Spread Props](https://svelte.dev/tutorial/spread-props))
+- Lit ([Discussion for support](https://github.com/lit/lit/issues/923), [PR](https://github.com/lit/lit/pull/1960))
+
+## Detailed Design
+
+A new LWC directive, `lwc:spread` will be introduced that can be used to spread the props to children elements.
+
+### Structure
+It accepts an object with keys as the property names and values.
+
+This object is merged with props passed in the template (essentially `Object.assign`).
+
+There is one difference however: the set of properties passed in the template doesn't change but it changes in `lwc:spread` i.e.
+a property can't be present in one render and absent in the next render if it is defined in the template. With
+`lwc:spread` though, it can be present in one render and not be present in the next. See example below.
+
+**Example**
+
+```html
+<template>
+    <x-child name={name}></x-child> <!-- in first render -->
+    <x-child></x-child> <!-- in second render, name is missing. This is not possible with template compiler -->
+</template>
+```
+
+```html
+<template>
+    <x-child lwc:spread={props}>
+</template>
+```
+
+```js
+export default class App extends LightningElement {
+    handleClick() {
+        this.props = showName ? {name: 'lwc'} : {}; // in one render we're passing `name` prop, but in another one there's no `name` property.
+    }
+
+}
+```
+
+In cases where a property from earlier render is missing, we have two options:
+1. Explicity set it to `undefined` or `null`.
+2. Leave it untouched. Meaning we just assign the properties we get in the current render without touching any of the earlier set properties.
+
+We'll go ahead with Option 1. We'll explicity set them to `undefined`. The advantage of this is that it guarantees that at any point
+the properties on an element are the ones that are present in the template and the ones present in `lwc:spread` in that render.
+
+An interesting side effect of the above behavior can be observed when properties have default values:
```

Comment:
also I didn't really understood why default values are important? I don't think they are. Default values in HTML are defined as initialization only, and they have side effects once, the first time they are set. I don't think we are talking about the same thing here... just confused.